src/dev/i18n_tools/extractors/call_expt.ts

```diff
@@ -0,0 +1,403 @@
+import ts from 'typescript';
+type TypeScript = typeof ts;
+
+const MESSAGE_DESC_KEYS: Array<keyof MessageDescriptor> = [
+  'id',
+  'defaultMessage',
+  'description',
+  'ignoreTag',
+];
+import type { MessageDescriptor } from '../types';
+
+export interface ExtractorOpts extends Opts {
+  onMsgWithValuesExtracted: (filePath: string, msgs: MessageDescriptor[]) => void;
+}
+
+export { MessageDescriptor };
+
+/**
+ * Check if node is `i18n.translate` node
+ * @param node
+ * @param sf
+ */
+function isMemberMethodI18nTranslateCall(typescript: TypeScript, node: ts.CallExpression) {
+  const fnNames = new Set(['translate']);
+  const method = node.expression;
+
+  // Handle foo.formatMessage()
+  if (typescript.isPropertyAccessExpression(method)) {
+    return fnNames.has(method.name.text);
+  }
+
+  // Handle formatMessage()
+  return typescript.isIdentifier(method) && fnNames.has(method.text);
+}
+
+/**
+ * Check if node is `foo.bar.formatMessage` node
+ * @param node
+ * @param sf
+ */
+function isMemberMethodFormatMessageCall(typescript: TypeScript, node: ts.CallExpression) {
+  const fnNames = new Set(['formatMessage']);
+  const method = node.expression;
+
+  // Handle foo.formatMessage()
+  if (typescript.isPropertyAccessExpression(method)) {
+    return fnNames.has(method.name.text);
+  }
+
+  // Handle formatMessage()
+  return typescript.isIdentifier(method) && fnNames.has(method.text);
+}
+
+function setAttributesInObject(
+  typescript: TypeScript,
+  factory: ts.NodeFactory,
+  node: ts.ObjectLiteralExpression,
+  msg: MessageDescriptor,
+  ast?: boolean
+) {
+  const newProps = [
+    factory.createPropertyAssignment('id', factory.createStringLiteral(msg.id)),
+    factory.createPropertyAssignment('ignoreTag', factory.createTrue()),
+  ];
+
+  for (const prop of node.properties) {
+    if (
+      typescript.isPropertyAssignment(prop) &&
+      typescript.isIdentifier(prop.name) &&
+      MESSAGE_DESC_KEYS.includes(prop.name.text as keyof MessageDescriptor)
+    ) {
+      newProps.push(prop);
+      continue;
+    }
+    if (typescript.isPropertyAssignment(prop)) {
+      newProps.push(prop);
+    }
+  }
+  return factory.createObjectLiteralExpression(factory.createNodeArray(newProps));
+}
+
+export function extractMessagesFromCallExpression(
+  typescript: TypeScript,
+  factory: ts.NodeFactory,
+  node: ts.CallExpression,
+  opts: ExtractorOpts = { onMsgWithValuesExtracted: () => {} },
+  sf: ts.SourceFile
+): ts.VisitResult<ts.CallExpression> {
+  if (isMemberMethodFormatMessageCall(typescript, node)) {
+    const [descriptorsObj, valuesObj, ...restArgs] = node.arguments;
+
+    if (!valuesObj) {
+      return node;
+    }
+
+    const msg = extractMessageDescriptor(typescript, descriptorsObj, opts, sf);
+    const valuesKeys = getObjectKeys(typescript, valuesObj);
+
+    const hasValuesObject = valuesKeys.length;
+    if (restArgs.length) {
+      // TODO: support parsing ignoreTag from 3rd argument.
```

Comment:
Is this part of this PR or future work? If the later, can we move it to an issue so that it can be tracked?

Reply:
I just forgot removing this personal TODO while writing the code